rust/otel-arrow-rust/src/otlp/attributes/cbor.rs

```diff
@@ -1,96 +1,135 @@
-            ciborium::Value::Integer(int_val) => Some(Value::IntValue(
-                int_val
-                    .try_into()
-                    .context(error::InvalidSerializedIntAttributeValueSnafu)?,
-            )),
-            ciborium::Value::Array(array_vals) => Some(Value::try_from(array_vals)?),
-            ciborium::Value::Map(kv_vals) => Some(Value::try_from(kv_vals)?),
-            other => {
-                return error::UnsupportedSerializedAttributeValueSnafu { actual: other }.fail();
+fn proto_encode_cbor_value(value: &ciborium::Value, result_buf: &mut ProtoBuffer) -> Result<()> {
+    match value {
+        ciborium::Value::Null => {
+            // do nothing, it's an empty value
+        }
+        ciborium::Value::Bool(bool_val) => {
+            result_buf.encode_field_tag(ANY_VALUE_BOOL_VALUE, wire_types::VARINT);
+            result_buf.encode_varint(*bool_val as u64);
+        }
+        ciborium::Value::Bytes(bytes_val) => {
+            result_buf.encode_bytes(ANY_VALUE_BYTES_VALUE, bytes_val);
+        }
+        ciborium::Value::Float(float_val) => {
+            result_buf.encode_field_tag(ANY_VALUE_DOUBLE_VALUE, wire_types::FIXED64);
+            result_buf.extend_from_slice(&float_val.to_le_bytes());
+        }
+        ciborium::Value::Text(str_val) => {
+            result_buf.encode_string(ANY_VALUE_STRING_VALUE, str_val);
+        }
+        ciborium::Value::Integer(int_val) => {
+            let int_val: u64 = (*int_val)
+                .try_into()
+                .context(error::InvalidSerializedIntAttributeValueSnafu)?;
+            result_buf.encode_field_tag(ANY_VALUE_INT_VALUE, wire_types::VARINT);
+            result_buf.encode_varint(int_val);
+        }
+        ciborium::Value::Array(list_val) => {
+            proto_encode_len_delimited_unknown_size!(
+                ANY_VALUE_ARRAY_VALUE,
+                proto_encode_array_values(list_val, result_buf)?,
+                result_buf
+            );
+        }
+        ciborium::Value::Map(kv_list_val) => {
+            proto_encode_len_delimited_unknown_size!(
+                ANY_VALUE_KVLIST_VALUE,
+                proto_encode_cbor_kv_list(kv_list_val, result_buf)?,
+                result_buf
+            );
+        }
+        other => {
+            return error::UnsupportedSerializedAttributeValueSnafu {
+                actual: other.clone(),
             }
-        };
-
-        Ok(Self(val))
+            .fail();
+        }
     }
+
+    Ok(())
 }
 
-/// Converts an array of cbor values into the ArrayValue variant of Value
-impl TryFrom<Vec<ciborium::Value>> for Value {
-    type Error = Error;
+fn proto_encode_array_values(
+    list_val: &[ciborium::Value],
+    result_buf: &mut ProtoBuffer,
+) -> Result<()> {
+    for v in list_val {
+        proto_encode_len_delimited_unknown_size!(
+            ARRAY_VALUE_VALUES,
+            proto_encode_cbor_value(v, result_buf)?,
```

Comment:
We should consider enforcing some depth limit for the recursion. Otherwise, it might be exploited by a bad actor. A deeply nested malicious payload may lead to StackOverflow/ DoS.

Reply:
Something to check for the OTLP proto bytes -> OTAP conversion as well:

https://github.com/open-telemetry/otel-arrow/blob/eaf81aabee7eaa80d4ffe78a59d9bfd0922345d0/rust/otap-dataflow/crates/otap/src/encoder/cbor.rs#L72-L89